### PR TITLE
net: ip: net_pkt: Fix net_pkt_pull data buffer corruption

### LINKED ISSUE #22307: net: ip: net_pkt_pull(): packet corruption when using CONFIG_NET_BUF_DATA_SIZE larger than 256
**Describe the bug**
`CONFIG_NET_BUF_DATA_SIZE` has a default of 128.
When using `CONFIG_NET_BUF_FIXED_DATA_SIZE=y` and `CONFIG_NET_BUF_DATA_SIZE` > 256 (e.g. 512), `net_pkt_pull()` corrupts the packet, as it calculates the lengths using `u8_t`.

This can be reproduced when using ipv6 with packet fragmentation. The packet reassembling code uses `net_pkt_pull()` on received packets.
You can send large ipv6 pings (e.g. `ping6 -s 2000 192.0.2.2`) to see the corrupted data

**Impact**
This is a showstopper when working with fragmented IPV6 packets.

**Environment (please complete the following information):**
 - OS: Linux
 - Toolchain: Zephyr SDK
 - Version: current master

**Additional context**

The solution is to change the type of `left, rem` from `u8_t` to `size_t`:
```
diff --git a/subsys/net/ip/net_pkt.c b/subsys/net/ip/net_pkt.c
index dfd2596c8d..8ceb13ffea 100644
--- a/subsys/net/ip/net_pkt.c
+++ b/subsys/net/ip/net_pkt.c
@@ -1848,7 +1848,7 @@ int net_pkt_pull(struct net_pkt *pkt, size_t length)
        net_pkt_cursor_backup(pkt, &backup);
 
        while (length) {
-               u8_t left, rem;
+               size_t left, rem;
 
                pkt_cursor_advance(pkt, false);
```

### PR DESCRIPTION
**Pull request by a8961713**
Fixes #22307

### CODE REVIEW

**Review by tbursztyka — COMMENTED** [review]
looks like you inverted the commit messages and the actual patches

**Comment by a8961713** [comment]
> looks like you inverted the commit messages and the actual patches

You are correct. I'll resubmit.

File: subsys/net/ip/net_pkt.c [hunk 1]
```diff
@@ -1848,7 +1848,7 @@ int net_pkt_pull(struct net_pkt *pkt, size_t length)
 	net_pkt_cursor_backup(pkt, &backup);
```
Comment (jukkar) [inline subsys/net/ip/net_pkt.c]: No need to duplicate the "net" prefix so many times in commit subject, for the first commit you can just say `net: pkt: Fixed packet buffer corruption in net_pkt_pull()`

File: tests/net/net_pkt/src/main.c [hunk 2]
```diff
@@ -629,6 +629,49 @@ void test_net_pkt_copy(void)
 		     "Pkt not properly unreferenced");
```
Comment (jukkar) [inline tests/net/net_pkt/src/main.c]: for the second commit, please change the subject to `tests: net: pkt: Added a test case for net_pkt_pull()`

**Review by jukkar — CHANGES_REQUESTED** [review]
Looks good, some minor nits.

**Review by jukkar — APPROVED** [review]

**Review by tbursztyka — APPROVED** [review]

File: tests/net/net_pkt/src/main.c [hunk 3]
```diff
@@ -629,6 +629,49 @@ void test_net_pkt_copy(void)
...
+		      PULL_TEST_PKT_DATA_SIZE - PULL_AMOUNT,
+		      "Pull failed to set new size");
+	zassert_true(net_pkt_read(dummy_pkt,
+				  pkt_data_readback,
+				  PULL_TEST_PKT_DATA_SIZE - PULL_AMOUNT) == 0,
+		     "Read packet failed");
+	zassert_mem_equal(pkt_data_readback,
+			  &pkt_data[PULL_AMOUNT],
+			  PULL_TEST_PKT_DATA_SIZE - PULL_AMOUNT,
+			  "Pakcet data changed");
```
Comment (pfalcon) [inline tests/net/net_pkt/src/main.c]: Typo: "Pakcet".

**Review by pfalcon — CHANGES_REQUESTED** [review]
Typo.

File: tests/net/net_pkt/src/main.c (hunk 3 above)
Comment (a8961713) [reply tests/net/net_pkt/src/main.c]: fixed, resubmitted.

**Review by pfalcon — APPROVED** [review]

**Pull request merged**